File: src/main/java/net/vidageek/games/regex/task/MassNegateAndMatch.java
package net.vidageek.games.regex.task;

import static net.vidageek.games.regex.task.MatcherTargets.from;

import java.io.InputStream;
import java.util.Scanner;

import net.vidageek.games.task.JudgedTask;
import net.vidageek.games.task.Task;

public class MassNegateAndMatch implements Task {

	private final String challenge;
	private final NegateAndMatch matcher;

	public MassNegateAndMatch(final String fileName, final String challenge) {
		this.challenge = challenge;
		InputStream stream = MassNegateAndMatch.class.getResourceAsStream("/mass/" + fileName + ".match");
		String[] mustMatch = new String[] {};
		if (stream != null) {
			mustMatch = new Scanner(stream).useDelimiter("$$").next().split("\n");
		}
		stream = MassNegateAndMatch.class.getResourceAsStream("/mass/" + fileName + ".not");
		String[] cantMatch = new String[] {};
		if (stream != null) {
			cantMatch = new Scanner(stream).useDelimiter("$$").next().split("\n");
		}
		matcher = new NegateAndMatch(from(cantMatch), from(mustMatch));
	}

	public JudgedTask judge(final String challenge) {
		return matcher.judge(challenge);
	}

	public String getChallenge() {
		return challenge;
	}

}

File: src/main/scala/net/vidageek/games/regex/task/MassNegateAndMatch.scala
package net.vidageek.games.regex.task

import java.util.Scanner
import net.vidageek.games.task.JudgedTask
import net.vidageek.games.task.Task

class MassNegateAndMatch(challenge: String, matcher: NegateAndMatch) extends Task {
  def judge(challenge: String): JudgedTask = matcher.judge(challenge)
  def getChallenge() = challenge
}

object MassNegateAndMatch {
  import MatcherTargets._
  
  def apply(fileName: String, challenge: String): MassNegateAndMatch = {
    val mustMatch = arrayFromFile("/mass/" + fileName + ".match")
    val cantMatch = arrayFromFile("/mass/" + fileName + ".not")
    val matcher = new NegateAndMatch(from(cantMatch), from(mustMatch))
    
    new MassNegateAndMatch(fileName, matcher)
  }
  
  private def arrayFromFile(filePath: String) = {
    val inputStream = classOf[MassNegateAndMatch].getResourceAsStream(filePath)
    if (inputStream == null) 
      Array[String]()
    else 
      new Scanner(inputStream).useDelimiter("$$").next().split("\n")
  }
}